Make no_log=False override the no_log warnings.

nnhanndakara opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
As AnsibleModule._log_invocation is currently implemented, any parameter
with a name that matches PASSWORD_MATCH triggers the no_log warning as a
precaution against parameters that may contain sensitive data, but have not
been marked as sensitive by the module author.

This patch would allow module authors to explicitly mark the aforementioned
parameters as not sensitive thereby bypassing an erroneous warning message,
while still catching parameters which have not been marked at all by the
author.

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
Fixes #49465
Fixes #64656

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
lib/ansible/module_utils/basic.py

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->
See #49465 and #64656.

<!--- Paste verbatim command output below, e.g. before and after your change -->
Before (with pass_interval set to no_log=False):
```paste below
% ansible -m mac_domain_membership -a "dns_domain_name=x domain_admin_password=x domain_admin_user=x" --check xxxxxxxx-xxxxxx
[WARNING]: Module did not set no_log for pass_interval

xxxxxxxx-xxxxxx | SUCCESS => {
*snip*
}
```
After (with pass_interval set to no_log=False):
```paste below
% ansible -m mac_domain_membership -a "dns_domain_name=x domain_admin_password=x domain_admin_user=x" --check xxxxxxxx-xxxxxx
xxxxxxxx-xxxxxx | SUCCESS => {
*snip*
}
```

nnhanndakara committed:
Make no_log=False override the no_log warnings.

As AnsibleModule._log_invocation is currently implemented, any parameter
with a name that matches PASSWORD_MATCH triggers the no_log warning as a
precaution against parameters that may contain sensitive data, but have not
been marked as sensitive by the module author.

This patch would allow module authors to explicitly mark the aforementioned
parameters as not sensitive thereby bypassing an erroneous warning message,
while still catching parameters which have not been marked at all by the
author.

Fixes: #49465 #64656

kaorihinata commented:
It's worth mentioning at this point that the way that `AnsibleModule._log_invocation` makes decisions about `no_log` parameters does not match how `AnsibleModule._handle_no_log_values` (and underlying `list_no_log_values()`) handles them. This is true both before and after this change, meaning that sanitizing output for parameters that match `PASSWORD_MATCH` is still just as broken as it always has been. If a more complete solution is desired, then by all means, we can close this pull request.

nnhanndakara committed:
Add changelog fragment for PR 64733.

bcoca reviewed (approved):
The password matching will never be perfect, this is mostly meant as a warning/fallback, as such this change does do the right thing giving the developer control. 
The warning will work in 'absence' of information and only trigger when there is lack of intended behaviour. Good change.

kaorihinata commented:
That's fine. I just figured I would mention it.

I was going to go ahead and fix the test failure by updating the core user module to take advantage of this change. It looks like the extra conditions that had been stapled onto the `PASSWORD_MATCH` check were added specifically for situations like this, but having module developers explicitly acknowledge it would replace such behavior. Should I go ahead and push it?

kaorihinata commented:
As far as I'm aware all tests related to the changed code are passing. The one test that is failing is failing due to instability (azure_rm_networkinterface, in this case). I do not believe I have the ability to rerun the test from here, and the test completed successfully when it was rerun with maximum verbosity.

Akasurde commented:
@bcoca Are we good at merging this? Thanks,

samdoran commented on changelogs/fragments/64733-make-no_log-false-override-no_log-warnings.yml:
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+  - "core: Make no_log=False override the no_log warnings - fixes #49465 and #64656"
```
```suggestion
  - "make ``no_log=False`` on a module option silence the ``no_log`` warning (https://github.com/ansible/ansible/issues/49465 https://github.com/ansible/ansible/issues/64656)"
```

samdoran commented on lib/ansible/modules/system/user.py:
```diff
@@ -2859,9 +2859,9 @@ def main():
             ssh_key_comment=dict(type='str', default=ssh_defaults['comment']),
             ssh_key_passphrase=dict(type='str', no_log=True),
-            update_password=dict(type='str', default='always', choices=['always', 'on_create']),
+            update_password=dict(type='str', default='always', choices=['always', 'on_create'], no_log=False),
```
These changes to the `user` module are good, but should be done in a separate PR.

nnhanndakara replied:
Agreed, but just to confirm, the only reason I made this change was that, as the current tests are written for the `user` module, the existence of any warning fails the tests, which fails CI. Is this okay?

samdoran replied:
We'll need to merge the user module PR before this one. Once that PR is merged, tests for this PR will pass.

nnhanndakara replied:
Pull request #66219 has been created. Is it okay to mark this as resolved?

Also, please let me know if anything needs to be done or redone. I'm not the most familiar with GitHub's tools yet and I'm already regretting not creating a branch for this pull request.

samdoran replied:
> Pull request #66219 has been created. Is it okay to mark this as resolved?

Yup.

> Also, please let me know if anything needs to be done or redone.

You can remove the changes to the `user` module from this PR.

> I'm not the most familiar with GitHub's tools yet and I'm already regretting not creating a branch for this pull request.

Creating new branches in your fork for a PR isn't required but it does make it easier to segment work.

samdoran reviewed (changes requested):
This is a very good change. Please add unit tests to the [existing argument spect tests](https://github.com/ansible/ansible/blob/devel/test/units/module_utils/basic/test_argument_spec.py) for this change.

samdoran commented:
We should also update the docs explaining this can be set to `True` _or_ `False` for a module option now. That will be very helpful information for module authors.

nnhanndakara committed:
Update changelogs/fragments/64733-make-no_log-false-override-no_log-warnings.yml

Co-Authored-By: Sam Doran <sdoran@redhat.com>

samdoran commented:
Re-running tests now that #66219 is merged.

samdoran commented:
Changelog and code change looks good. We still need unit tests and documentation updates. Let me know if you need help with that.

kaorihinata commented:
For the documentation, `docs/docsite/rst/dev_guide/developing_modules_documenting.rst` and `docs/docsite/rst/dev_guide/developing_program_flow_modules.rst` seem to be the only documents that will be affected. Does this sound correct?

I noticed that the C# module for Powershell implements `no_log` in some form, but that it seems to operate differently than the Python module.

Regarding the tests, should I just add a VALID_SPECS case for each value of no_log?

kaorihinata commented:
Now that I look at it more closely, I don't think I'll be able to get away with that.

Additionally, it would appear that AWS is having a moment.

bcoca commented:
checked with windows folk, they don't have a matching feature so no need to update the powershell common code.

samdoran commented:
> For the documentation, docs/docsite/rst/dev_guide/developing_modules_documenting.rst and docs/docsite/rst/dev_guide/developing_program_flow_modules.rst seem to be the only documents that will be affected. Does this sound correct?

Yes.

> I noticed that the C# module for Powershell implements no_log in some form, but that it seems to operate differently than the Python module.

The C# implementation may already match the behavior in this PR. I don't see anything in `Ansible.Basic.cs` that is warning based on a regexp match of the module parameter names. Maybe @jborean93 or @nitzmahone can add further clarification.

> Regarding the tests, should I just add a VALID_SPECS case for each value of no_log?

It's probably easiest to add new test(s) specifically for testing the various `no_log` settings. Here's a basic example:

```python
@pytest.mark.parametrize('stdin', [{'ok_password': 'notsecret'}], indirect=['stdin'])
def test_no_log_false_no_warning(stdin, capfd):
    spec = {
        'ok_password': {'no_log': False},
    }
    am = basic.AnsibleModule(spec)
    assert len(am._warnings) == 0
```

samdoran commented:
> Additionally, it would appear that AWS is having a moment.

That happens a lot. 😄

bcoca commented:
@samdoran read my post, i asked them already during windows irc meeting

kaorihinata commented:
> That happens a lot. 😄

Serves me right for watching the build status. I'll have the documentation in some time this evening.

nnhanndakara committed:
Update docs to include usage of no_log=False.

Updates the Sphinx documentation to include directions for the appropriate
use of no_log=False to avoid having Ansible automatically remove an argument
value from logs and output.

samdoran commented on docs/docsite/rst/dev_guide/developing_modules_documenting.rst:
```diff
@@ -119,7 +119,8 @@ After the shebang, the UTF-8 coding, the copyright line, the license, and the ``
     * Verify that arguments in doc and module spec dict are identical.
-    * For password / secret arguments no_log=True should be set.
+    * For password / secret arguments ``no_log=True`` should be set.
+    * For arguments that are redacted automatically (e.g. "password_length"), but are not password / secret arguments, set ``no_log=False``.
```
```suggestion
    * For arguments that seem to contain sensitive information but **do not** contain secrets, such as "password_length", set ``no_log=False`` to disable the warning message.
```

samdoran replied:
We avoid using Latin abbreviations like 'e.g.' that are less familiar to those for whom English is not their native language.

samdoran commented on docs/docsite/rst/dev_guide/developing_program_flow_modules.rst:
```diff
@@ -615,7 +615,10 @@ required
 """"""
 
-``no_log`` indicates that the value of the argument should not be logged or displayed.
+``no_log`` accepts a boolean, either ``True`` or ``False`` that indicates explicitly whether or not the argument value should appear redacted in logs and output.
```
```suggestion
``no_log`` accepts a boolean, either ``True`` or ``False``, that indicates explicitly whether or not the argument value should be masked in logs and output.
```

samdoran commented on docs/docsite/rst/dev_guide/developing_program_flow_modules.rst:
```diff
@@ -615,7 +615,10 @@ required
+``no_log`` accepts a boolean, either ``True`` or ``False`` that indicates explicitly whether or not the argument value should appear redacted in logs and output.
+
+.. note::
+   In the absence of ``no_log``, the argument value will only appear redacted in logs and output if the argument name appears to indicate that the argument value is a password or passphrase (e.g. "admin_password"). To ensure that an argument value always appears in logs and output, and avoid misleading warnings about ``no_log``, set ``no_log`` explicitly to ``False``.
```
```suggestion
   In the absence of ``no_log``, the argument value will only be masked in logs and output if the argument name appears to indicate that the argument value is a password or passphrase (such as "admin_password"). To ensure that an argument value always appears in logs and output, and to disable the warning about ``no_log``, set ``no_log`` explicitly to ``False``.
```

samdoran reviewed (changes requested):
Some minor changes to the wording in the documentation.

nnhanndakara committed:
Update docs/docsite/rst/dev_guide/developing_program_flow_modules.rst

Co-Authored-By: Sam Doran <sdoran@redhat.com>

nnhanndakara committed:
Update docs/docsite/rst/dev_guide/developing_program_flow_modules.rst

Co-Authored-By: Sam Doran <sdoran@redhat.com>

nnhanndakara committed:
Update docs/docsite/rst/dev_guide/developing_modules_documenting.rst

Co-Authored-By: Sam Doran <sdoran@redhat.com>

kaorihinata commented:
Masked! That's much better. I don't know what it was last night but I was drawing a blank about what to put there. I'll remember the latin policy in the future.

Regarding the comma, I had assumed a comma would be appropriate there but the description for `required` above `no_log` excludes it. Should one be insert there as well for consistency, or should that be reserved for another pull request specifically targeting documentation corrections?

samdoran commented:
We don't need to worry about the comma. It's ok how it is for now.

nnhanndakara committed:
Add tests for various no_log states.

Adds tests for various no_log states including True, False, and None (as
extracted by AnsibleModule._log_invocation) when applied to an argument with
a name that matches PASSWORD_MATCH.

samdoran commented on docs/docsite/rst/dev_guide/developing_program_flow_modules.rst:
```diff
@@ -615,7 +615,10 @@ required
+``no_log`` accepts a boolean, either ``True`` or ``False``, that indicates explicitly whether or not the argument value should be masked in logs and output.
+
+.. note::
+   In the absence of ``no_log``, the argument value will only be masked in logs and output if the argument name appears to indicate that the argument value is a password or passphrase (such as "admin_password"). To ensure that an argument value always appears in logs and output, and to disable the warning about ``no_log``, set ``no_log`` explicitly to ``False``.
```
Actually, I don't think this is accurate. The value is not masked if the parameter name contains something like 'password' — only a warning is issued.
```suggestion
   In the absence of ``no_log``, if the parameter name appears to indicate that the argument value is a password or passphrase (such as "admin_password"), a warning will be shown and the value will be masked in logs but **not** output. To disable the warning and masking for parameters that do not contain sensitive information, set ``no_log`` to ``False``.
```

nnhanndakara replied:
I'm not sure if it's that simple. `log_args[param]` does get initialized to `"NOT_LOGGING_PASSWORD"`, but I haven't finished looking at where that value ends up. Would you happen to know off-hand?

Edit: Also, thanks. I'd forgotten about this oddity. I believe it's related to how `_log_invocation` works versus how `_handle_no_log_values` works. A parameter found in `_log_invocation` never makes it into `no_log_values`.

samdoran replied:
It gets passed to the `log()` method, but it doesn't actually mask anything since `remove_values()` is using the values in `self.list_no_log_values`.

This is probably a bug because I bet the intention was to mask those values if they looked like that contained sensitive information.

samdoran replied:
Oh, nevermind. It _does_ actually get masked in the log message, just not in the module invocation.

```
syslog_msg='Invoked with password_length=NOT_LOGGING_PASSWOR...t_elements
```

nnhanndakara replied:
Ahh. That's what I was trying to test. Thanks.

samdoran replied:
So the value is masked in the syslog, but not the Ansible log/output.

nnhanndakara replied:
You could technically add the value to `no_log_values` on the module object with each invocation of `_log_invocation` (since it's a `set`) and eat the cost of the call, but that would  result in an inconsistency between the resulting messages, I believe. `"NOT_LOGGING_PASSWORD"` in syslog, and `"NO_LOGGING_PARAMETER"` in the output.

nnhanndakara replied:
For the purposes of this PR, should we be documenting the intended behavior, or the current behavior, including the bug?

samdoran replied:
I wonder if that's an omission. I will ask around. It's simple to also mask that value in output.

```patch
diff --git lib/ansible/module_utils/basic.py lib/ansible/module_utils/basic.py
index a96ce6aa5d..14a22bdf59 100644
--- lib/ansible/module_utils/basic.py
+++ lib/ansible/module_utils/basic.py
@@ -1948,6 +1948,7 @@ class AnsibleModule(object):
             elif PASSWORD_MATCH.search(param) and arg_opts.get('type', 'str') != 'bool' and not arg_opts.get('choices', False):
                 # skip boolean and enums as they are about 'password' state
                 log_args[param] = 'NOT_LOGGING_PASSWORD'
+                self.no_log_values.add(self.params[param])
                 self.warn('Module did not set no_log for %s' % param)
             else:
                 param_val = self.params[param]

```

samdoran replied:
We discussed this internally, and adding the suspect value to `self.no_log_values` could potentially cause issues by masking a lot of output strings that don't need to be masked.

nnhanndakara replied:
Potentially. If the consensus is that the responsibility placed on module authors also needs to come with some safeguards then I can understand the logic behind it. Should this be documented that way in the documentation then?

samdoran replied:
Yes. It is up to the module author to mark fields as `no_log`.

nnhanndakara committed:
PEP8 cleanup for test_argument_spec.

Added extra spacing as recommended by pycodestyles. This should satsify the
sanity/1 test in CI.

samdoran commented:
We have integration tests for `no_log` [here](https://github.com/ansible/ansible/tree/devel/test/integration/targets/no_log). You should be able to add to these tests for this new behavior.

nnhanndakara committed:
Update docs/docsite/rst/dev_guide/developing_program_flow_modules.rst

Co-Authored-By: Sam Doran <sdoran@redhat.com>

samdoran commented:
This is listed as a bug fix, but it is also a change in behavior. I'll ask for other opinions about whether or not it should be backported.

kaorihinata commented:
Not all modules are tested like `user`. There are at least a few modules currently in the tree that will emit a warning with this change, but they should be pretty easy to spot in CI output.

samdoran commented:
The consensus on backporting this is not strong either way, so go ahead and open a [backport PR](https://docs.ansible.com/ansible/latest/community/development_process.html#backporting-merged-prs).

kaorihinata commented:
Should be doable. I'm moving house so it may not happen immediately, but I'll try to get it thrown together between bouts of filling boxes.
